File: Backend/src/main/java/in/osop/messaging_platform/service/WebhookService.java
package in.osop.messaging_platform.service;

import in.osop.messaging_platform.dto.WebhookEndpointDto;
import in.osop.messaging_platform.model.WebhookEndpoint;

import java.util.List;
import java.util.Map;

public interface WebhookService {
    
    /**
     * Create a new webhook endpoint
     * @param tenantId The tenant ID
     * @param webhookDto The webhook data
     * @return The created webhook endpoint
     */
    WebhookEndpointDto createWebhook(Long tenantId, WebhookEndpointDto webhookDto);
    
    /**
     * Get all webhook endpoints for a tenant
     * @param tenantId The tenant ID
     * @return List of webhook endpoints
     */
    List<WebhookEndpointDto> getAllWebhooks(Long tenantId);
    
    /**
     * Get a webhook endpoint by ID
     * @param tenantId The tenant ID
     * @param webhookId The webhook ID
     * @return The webhook endpoint
     */
    WebhookEndpointDto getWebhookById(Long tenantId, Long webhookId);
    
    /**
     * Update a webhook endpoint
     * @param tenantId The tenant ID
     * @param webhookId The webhook ID
     * @param webhookDto The updated webhook data
     * @return The updated webhook endpoint
     */
    WebhookEndpointDto updateWebhook(Long tenantId, Long webhookId, WebhookEndpointDto webhookDto);
    
    /**
     * Delete a webhook endpoint
     * @param tenantId The tenant ID
     * @param webhookId The webhook ID
     */
    void deleteWebhook(Long tenantId, Long webhookId);
    
    /**
     * Toggle webhook status (enable/disable)
     * @param tenantId The tenant ID
     * @param webhookId The webhook ID
     * @param enabled The new enabled status
     * @return The updated webhook endpoint
     */
    WebhookEndpointDto toggleWebhookStatus(Long tenantId, Long webhookId, boolean enabled);
    
    /**
     * Get a list of all available webhook event types
     * @return List of event type names
     */
    List<String> getAvailableEvents();
    
    /**
     * Send a webhook notification for an event
     * @param tenantId The tenant ID
     * @param eventType The event type
     * @param payload The event payload
     */
    void sendWebhookNotification(Long tenantId, String eventType, Map<String, Object> payload);
    
    /**
     * Get webhook endpoints that are subscribed to a specific event
     * @param tenantId The tenant ID
     * @param eventType The event type
     * @return List of webhook endpoints
     */
    List<WebhookEndpoint> getWebhooksForEvent(Long tenantId, String eventType);
}

File: frontend/src/api/webhookService.ts
import { apiClient } from './client';
import { AxiosResponse } from 'axios';

export interface WebhookEndpoint {
  id: number;
  url: string;
  name: string;
  description?: string;
  events: string[];
  secretKey?: string;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface WebhookEndpointRequest {
  url: string;
  name: string;
  description?: string;
  events: string[];
  secretKey?: string;
  enabled?: boolean;
}

export interface ApiResponse<T> {
  success: boolean;
  message?: string;
  data?: T;
}

export const webhookService = {
  // Create a new webhook endpoint
  async createWebhook(request: WebhookEndpointRequest): Promise<ApiResponse<WebhookEndpoint>> {
    try {
      const response: AxiosResponse<WebhookEndpoint> = await apiClient.post('/webhooks', request);
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to create webhook endpoint'
      };
    }
  },

  // Get all webhook endpoints
  async getAllWebhooks(): Promise<ApiResponse<WebhookEndpoint[]>> {
    try {
      const response: AxiosResponse<WebhookEndpoint[]> = await apiClient.get('/webhooks');
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to fetch webhook endpoints'
      };
    }
  },

  // Get webhook endpoint by ID
  async getWebhookById(id: number): Promise<ApiResponse<WebhookEndpoint>> {
    try {
      const response: AxiosResponse<WebhookEndpoint> = await apiClient.get(`/webhooks/${id}`);
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to fetch webhook endpoint'
      };
    }
  },

  // Update webhook endpoint
  async updateWebhook(id: number, request: WebhookEndpointRequest): Promise<ApiResponse<WebhookEndpoint>> {
    try {
      const response: AxiosResponse<WebhookEndpoint> = await apiClient.put(`/webhooks/${id}`, request);
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to update webhook endpoint'
      };
    }
  },

  // Delete webhook endpoint
  async deleteWebhook(id: number): Promise<ApiResponse<void>> {
    try {
      await apiClient.delete(`/webhooks/${id}`);
      return {
        success: true
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to delete webhook endpoint'
      };
    }
  },

  // Toggle webhook endpoint status (enable/disable)
  async toggleWebhookStatus(id: number, enabled: boolean): Promise<ApiResponse<WebhookEndpoint>> {
    try {
      const response: AxiosResponse<WebhookEndpoint> = await apiClient.post(`/webhooks/${id}/toggle`, { enabled });
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to update webhook status'
      };
    }
  },

  // Get available webhook events
  async getAvailableEvents(): Promise<ApiResponse<string[]>> {
    try {
      const response: AxiosResponse<string[]> = await apiClient.get('/webhooks/events');
      return {
        success: true,
        data: response.data
      };
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to fetch available events'
      };
    }
  }
};
